File: traditional_comm/src/codecs.js
const zlib = require('node:zlib');

function assertBuffer(value, label) {
  if (!Buffer.isBuffer(value)) {
    throw new TypeError(`${label} must be a Buffer`);
  }
}

function encodePayload(kind, input) {
  assertBuffer(input, 'input');

  switch (kind) {
    case 'text':
      return {
        codec: 'utf8',
        container: 'txt',
        payload: Buffer.from(input),
      };
    case 'image':
      return {
        codec: 'ppm-deflate',
        container: 'ppm',
        payload: zlib.deflateSync(input, { level: 6 }),
      };
    case 'video':
      return {
        codec: 'tvid-deflate',
        container: 'tvid',
        payload: zlib.deflateSync(input, { level: 6 }),
      };
    default:
      throw new Error(`Unsupported media kind: ${kind}`);
  }
}

function decodePayload(kind, payload) {
  assertBuffer(payload, 'payload');

  switch (kind) {
    case 'text':
      return Buffer.from(payload);
    case 'image':
    case 'video':
      return zlib.inflateSync(payload);
    default:
      throw new Error(`Unsupported media kind: ${kind}`);
  }
}

function codecInfo(kind) {
  switch (kind) {
    case 'text':
      return { codec: 'utf8', container: 'txt', lossless: true };
    case 'image':
      return { codec: 'ppm-deflate', container: 'ppm', lossless: true };
    case 'video':
      return { codec: 'tvid-deflate', container: 'tvid', lossless: true };
    default:
      throw new Error(`Unsupported media kind: ${kind}`);
  }
}

module.exports = {
  encodePayload,
  decodePayload,
  codecInfo,
};

File: traditional_comm_py/traditional_comm/codecs.py
from __future__ import annotations

from dataclasses import dataclass
import zlib


@dataclass(frozen=True)
class EncodedPayload:
    codec: str
    container: str
    payload: bytes


def _check_bytes(value: bytes, label: str) -> None:
    if not isinstance(value, bytes):
        raise TypeError(f"{label} must be bytes")


def encode_payload(kind: str, data: bytes) -> EncodedPayload:
    _check_bytes(data, "data")
    if kind == "text":
        return EncodedPayload("utf8", "txt", data)
    if kind == "image":
        return EncodedPayload("ppm-deflate", "ppm", zlib.compress(data, level=6))
    if kind == "video":
        return EncodedPayload("tvid-deflate", "tvid", zlib.compress(data, level=6))
    raise ValueError(f"unsupported media kind: {kind}")


def decode_payload(kind: str, payload: bytes) -> bytes:
    _check_bytes(payload, "payload")
    if kind == "text":
        return bytes(payload)
    if kind in {"image", "video"}:
        return zlib.decompress(payload)
    raise ValueError(f"unsupported media kind: {kind}")


def codec_info(kind: str) -> dict:
    values = {
        "text": {"codec": "utf8", "container": "txt", "lossless": True},
        "image": {"codec": "ppm-deflate", "container": "ppm", "lossless": True},
        "video": {"codec": "tvid-deflate", "container": "tvid", "lossless": True},
    }
    try:
        return values[kind].copy()
    except KeyError as exc:
        raise ValueError(f"unsupported media kind: {kind}") from exc
